### ISSUE TITLE
How are packages actually found?

### ISSUE THREAD

**Issue by zerothi** [scikit-build/scikit-build-core#278]
From the doc https://scikit-build-core.readthedocs.io/en/latest/configuration.html#customizing-the-built-wheel it seems to me that `<package>` and `src/<package>` are always found.  
But it is not fully clear to me what `[tool.scikit-build]  packages = ["<package>"]` does.

What confuses me is its relation to `[tool.setuptools.packages.find]   where = [ ... ]`.

I also read your remark here https://github.com/scikit-build/scikit-build-core/issues/269#issuecomment-1505324069
which suggested that `[tool.scikit-build]  packages` only need the top-level package name. Or?

Basically I want to know which of these are needed:

```
- pack/
  - hello/
  - foo/
 ```
and whether I then need to do:
```
[tool.scikit-build]
packages = ["pack", "pack/hello", "pack/foo"]
```
or only `"pack"`?

If the latter then it has no resemblances to `setuptools` find method.

Also, a comment on namespace packages would be nice. But I'll test and comment here as well. :)

 https://github.com/scikit-build/scikit-build-core/issues/269

**Comment by henryiii** [comment]
Can’t describe fully ATM, but lookup hatchling, it’s designed after that. It’s only the top level package, subpackages (and data files) should be included automatically. I don’t remember if `wheel.exclude` is implemented yet.

Support for namespace packages is a bit sketchy at the moment. You can always fully control via CMake (but that’s not ideal for editable installs).

**Comment by henryiii** [comment]
:robot: _AI text below_ :robot:

Good questions — the key thing is that `wheel.packages` is **not** a search/find mechanism like setuptools' `tool.setuptools.packages.find.where`. Each entry names **one top-level package directory to copy verbatim**, and the last path component becomes the importable package name. Everything under it (subpackages, data files) is copied **recursively and automatically** — you never list subpackages.

For your layout it depends on what `pack/` is:

```
pack/
  hello/   # __init__.py
  foo/     # __init__.py
```

- If `pack/` is itself a package (has `pack/__init__.py`): use `packages = ["pack"]`. `hello` and `foo` come along automatically. **Not** `["pack", "pack/hello", "pack/foo"]`.
- If `pack/` is just a container dir (no `__init__.py`) and `hello`/`foo` are two *separate* top-level packages: list both — `packages = ["pack/hello", "pack/foo"]` — which install as top-level `hello` and `foo`. This is the closest analog to setuptools' `where = ["pack"]`, except you enumerate the packages rather than a search root.

So: only top-level package names, never subpackages — my earlier remark in #269 was correct.

Auto-detection (when `packages` is unset) only looks for a package matching your **project name** at `src/<name>`, `python/<name>`, then `<name>`. If you have multiple top-level packages, or the package lives elsewhere, you must list them explicitly.

**Namespace packages**: a dotted distribution name is handled as a PEP 420 layout (`ns.pkg` tries both `ns_pkg` and `ns/pkg`, where only the leaf needs an `__init__`), but for anything nontrivial it's best to list packages explicitly or drive the layout from CMake.

I've opened #1407 to clarify this in the docs.
